### ISSUE TITLE
Migrate Heightfield storage to wp.Texture2D + shape_source_ptr (mesh-style)

### ISSUE THREAD

**Issue by vreutskyy** [newton-physics/newton#2624]
## Motivation

``GeoType.HFIELD`` shapes currently carry their own three-table indirection on ``Model``:

- ``shape_heightfield_index: wp.array[wp.int32]`` (per-shape index, ``-1`` sentinel)
- ``heightfield_data: wp.array[HeightfieldData]`` (compact metadata table)
- ``heightfield_elevations: wp.array[wp.float32]`` (concatenated [0, 1] elevation grid)

Every raycast / collision / SDF kernel has to thread those arrays through and re-do the per-shape lookup. Meshes, by contrast, just stash a ``wp.Mesh`` handle in ``shape_source_ptr`` and dispatch through ``wp.mesh_query_ray`` -- one slot, one indirection. The asymmetry comes from there being no first-class ``wp.Heightfield`` in warp; ``HeightfieldData`` was rolled by hand to fill that gap.

PR #2560 had to add a factory + ``wp.static(enable_heightfields)`` gating because the per-thread ``HeightfieldData`` setup added measurable overhead (+12-63% per ``(pixel, shape)`` thread on non-HFIELD scenes per @eric-heiden's [benchmarks](https://github.com/newton-physics/newton/pull/2560#discussion_r2475336627)). That gating exists *only* because of the asymmetry -- meshes don't need it.

## Proposed design

Use ``wp.Texture2D`` (float32) as the storage for elevation values, store the texture handle in the existing ``shape_source_ptr`` slot, and put the heightfield extents in ``shape_size``. ``HeightfieldData``, ``shape_heightfield_index``, ``Model.heightfield_data``, ``Model.heightfield_elevations``, and ``Model.has_heightfields`` all go away.

### Separation of concerns

| Concern | Where it lives | Notes |
|---|---|---|
| Height function (the actual elevation grid) | ``wp.Texture2D`` content | Pre-normalized so ``sample(u, v) * z_scale = world_height``. ``min_z``/``max_z`` baked into the values. |
| Z-range scale (``z_scale``) | Per-asset float on ``Heightfield`` | Derived once at finalize from ``min_z``/``max_z``. |
| World half-extents (``hx``, ``hy``) | Per-asset floats on ``Heightfield`` | Already there; just propagated into ``shape_size`` at instance time. |
| Per-instance scale | ``shape_size`` (multiplied with the asset extents) | The only place per-instance state lives. |
| Asset handle | ``shape_source_ptr[shape_idx]`` (uint64) | Parallel to ``wp.Mesh.id``. |

``hx``/``hy`` do **not** participate in the texture normalization -- they're purely XY-footprint metadata. Only ``min_z``/``max_z`` shape the texture content.

### Texture content

Given a ``Heightfield`` with raw data ``raw[i, j]``, world-space ``min_z``, ``max_z``:

1. Normalize the raw data to ``[0, 1]`` (already done by the existing ``Heightfield`` constructor): ``data_norm[i, j] = (raw[i, j] - d_min) / (d_max - d_min)``.
2. Pick a per-asset ``z_scale = max(abs(min_z), abs(max_z))`` (degenerate ``min_z == max_z == 0`` case is fine -- ``z_scale = 0``, ``sample * 0 = 0``).
3. Texture stores: ``texture_value[i, j] = (min_z + data_norm[i, j] * (max_z - min_z)) / z_scale`` for ``z_scale \!= 0``, else zero.
4. By construction: ``texture_value * z_scale == world_height`` for every cell.

Examples:

| ``[min_z, max_z]`` | ``z_scale`` | Texture value range |
|---|---|---|
| ``[0, 1]`` | ``1`` | ``[0, 1]`` (raw normalized data) |
| ``[-1, 2]`` | ``2`` | ``[-0.5, 1]`` |
| ``[0, 5]`` | ``5`` | ``[0, 1]`` |
| ``[-3, -1]`` | ``3`` | ``[-1, -1/3]`` |
| flat at ``z = k`` | ``k`` | ``1`` everywhere |
| flat at ``z = 0`` | ``0`` | irrelevant (multiplied by 0) |

### shape_size encoding

```
shape_size[shape_idx] = (
    scale.x * heightfield.hx,
    scale.y * heightfield.hy,
    scale.z * heightfield.z_scale,
)
```

The instance's local-frame XY extent is ``[-shape_size.x, shape_size.x] x [-shape_size.y, shape_size.y]``. The texture's UV ``[0, 1]^2`` maps onto that. World height at any cell: ``sample(u, v) * shape_size.z``. Per-instance ``scale`` is applied **only** here, in one place.

### Lifetime

Mirrors how ``Mesh`` already manages ``wp.Mesh`` ([``types.py:976``](https://github.com/newton-physics/newton/blob/main/newton/_src/geometry/types.py#L976)):

```
Model
 -> shape_source : list[Any]                        # keeps Python refs alive
      -> Heightfield (user instance)
           -> self.texture : wp.Texture2D           # owns the warp object
                -> elevation buffer (wp.array)
 -> shape_source_ptr : wp.array[wp.uint64]          # raw handles for kernels
```

The ``Heightfield`` Python instance owns the ``wp.Texture2D`` via ``self.texture``. Builder appends the ``Heightfield`` to ``Model.shape_source`` (already does this). ``shape_source_ptr`` only carries the uint64 handle; lifetime is anchored in ``shape_source``.

Identical heightfield instances already share a single Python object via the builder's ``finalized_geos`` cache, so the same ``wp.Texture2D`` is reused across instances automatically.

### Kernel-side changes

``ray_intersect_geom`` HFIELD branch (and the analogous narrow-phase / SDF helpers) become roughly:

```python
# Already in local-frame from map_ray_to_local. Local extent comes from shape_size.
hx, hy, z_scale = size[0], size[1], size[2]
nrow = wp.texture_dim(tex_handle, 1)  # or whatever the warp builtin spelling is
ncol = wp.texture_dim(tex_handle, 0)
# 2D DDA across cells, sampling the texture for cell elevations:
h = wp.texture_sample(tex_handle, u, v) * z_scale
```

No more ``hfd.hx``, no ``hfd.min_z``, no ``elevation_data[base + r * ncol + c]``.

## Implementation steps

1. **Add ``Heightfield.finalize(device, requires_grad=False) -> wp.uint64``** (parallel to ``Mesh.finalize``):
   - Compute ``z_scale``; stash as ``self.z_scale``.
   - Build the texture content array.
   - ``self.texture = wp.Texture2D(...)``.
   - ``return self.texture._tex_handle``.
2. **Update ``ModelBuilder.finalize()`` HFIELD path** ([``builder.py:9967``](https://github.com/newton-physics/newton/blob/main/newton/_src/sim/builder.py#L9967)):
   - Replace the ``compact_heightfield_data`` / ``elevation_chunks`` build-up with ``shape_source_ptr[shape_idx] = heightfield.finalize(device=device)``.
   - Populate ``shape_size[shape_idx] = (scale.x * hf.hx, scale.y * hf.hy, scale.z * hf.z_scale)``.
   - Drop ``Model.shape_heightfield_index``, ``Model.heightfield_data``, ``Model.heightfield_elevations``, ``Model.has_heightfields`` declarations from ``model.py``.
3. **Update HFIELD AABB build path** ([``builder.py:9736``](https://github.com/newton-physics/newton/blob/main/newton/_src/sim/builder.py#L9736)) to read extents from ``shape_size`` instead of ``shape_src.hx * shape_scale[0]``.
4. **Migrate raycast** (``geometry/raycast.py``):
   - Drop ``_make_raycast_funcs`` factory + ``raycast_kernel_no_hfield`` / ``sensor_raycast_kernel_no_hfield`` lean variants.
   - Drop the ``shape_idx + shape_heightfield_index + heightfield_data + heightfield_elevations`` parameter trail from ``ray_intersect_geom``.
   - HFIELD branch reads texture handle from ``shape_source_ptr[shape_idx]``, dimensions from ``wp.texture_dim``, samples for elevations.
5. **Migrate ``utils/heightfield.py`` helpers** (``ray_intersect_heightfield_local``, ``sample_sdf_heightfield``, ``sample_sdf_grad_heightfield``, ``get_triangle_shape_from_heightfield``, ``heightfield_vs_convex_midphase``):
   - Replace ``hfd: HeightfieldData`` + ``elevation_data: wp.array`` parameters with ``tex_handle: wp.uint64`` + extents derived from ``shape_size``.
   - Replace ``elevation_data[base + r * hfd.ncol + c]`` with ``wp.texture_sample(tex_handle, ...)``.
6. **Migrate narrow-phase / SDF contact** (``geometry/sdf_contact.py``):
   - Drop ``get_edge_from_heightfield``'s ``hfd`` parameter, switch to texture-based sampling.
   - Update all callers in ``find_interesting_edges`` / ``do_edge_sdf_collision``.
7. **Update launch sites** (``picking.py``, ``sensor_raycast.py``):
   - Drop the per-Model ``has_heightfields`` variant selection -- single kernel pair after the migration.
   - Drop the three trailing array args from ``raycast_kernel`` / ``sensor_raycast_kernel`` launches.
8. **Update tests** (``test_raycast.py``, ``test_heightfield.py``):
   - Remove the ``shape_heightfield_index`` / ``heightfield_data`` / ``heightfield_elevations`` plumbing in test kernel inputs.
   - HFIELD test kernels just pass the ``wp.Texture2D._tex_handle`` like meshes pass ``wp.Mesh.id``.

## Compatibility with non-collision consumers

All non-collision paths read the user-facing ``Heightfield`` Python attributes (``hx``, ``hy``, ``min_z``, ``max_z``, ``nrow``, ``ncol``, ``data``) via ``model.shape_source[shape_idx]``. None of them touch ``HeightfieldData``, ``heightfield_elevations``, or even ``shape_source_ptr`` for HFIELD. The refactor preserves every existing ``Heightfield`` field; it only *adds* ``self.texture`` and ``self.z_scale``. So the following stay unchanged:

| Consumer | Access path | Status |
|---|---|---|
| MuJoCo solver ([``solver_mujoco.py:4331``](https://github.com/newton-physics/newton/blob/main/newton/_src/solvers/mujoco/solver_mujoco.py#L4331)) | ``model.shape_source[shape]`` | No change. ``mj_size = (hf.hx, hf.hy, hf.max_z - hf.min_z, eps)`` etc. still work. |
| MJCF importer ([``import_mjcf.py:975``](https://github.com/newton-physics/newton/blob/main/newton/_src/utils/import_mjcf.py#L975)) | ``Heightfield(...)`` constructor | No change. |
| GL viewer ([``viewer.py:867``](https://github.com/newton-physics/newton/blob/main/newton/_src/viewer/viewer.py#L867)) | ``model.shape_source[shape_idx]`` | No change. The render-mesh conversion reads ``data`` / ``hx`` / ``hy`` etc. as today. |

Aside: the MuJoCo path currently doesn't apply the per-instance ``shape_scale`` to the heightfield it builds in MuJoCo (it uses ``hf.hx`` / ``hf.hy`` directly). That's a *pre-existing* asymmetry and stays as-is post-refactor. Bringing MuJoCo collision into agreement with the new ``shape_size``-driven scale convention is a separate concern (would need either a per-instance MuJoCo hfield asset or a follow-up design tweak).

## Risks / open questions

- ``wp.Texture`` is **experimental** in warp 1.x ([source](https://github.com/NVIDIA/warp/blob/main/warp/_src/texture.py)). API is "subject to change without formal deprecation cycle." Migration locks us into watching that API.
- Need to confirm:
  - ``wp.texture_dim`` (or equivalent) available inside ``@wp.func`` / ``@wp.kernel``.
  - CPU sampling path is stable for our test matrix.
  - Float32 sampling does not implicitly normalize/clamp (we want raw values back).
  - Texture creation cost at ``finalize()`` is acceptable (vs the current ``wp.array`` upload).
- Behavior change to flag: under the new scheme, per-instance ``scale`` becomes a multiplier of the asset extents (a pure stretch). Document the migration in ``CHANGELOG.md`` even though the new semantics are arguably the obvious ones.

## Benefits

- True architectural parity with meshes -- one ``shape_source_ptr`` slot per shape, one warp-native handle per asset.
- Drops ~hundred lines of HFIELD plumbing across ``builder.py``, ``model.py``, ``raycast.py``, ``sdf_contact.py``, ``utils/heightfield.py``, ``picking.py``, ``sensor_raycast.py``.
- Eliminates the perf gating in #2560 by construction (no HFIELD-specific kernel args -> nothing to strip).
- Per-instance scale becomes free: it's just a multiplication when populating ``shape_size``.
- Texture cache can speed up DDA traversal for spatially coherent rays.

## Out of scope for this issue

- Whether to gate the migration behind a ``wp.Texture``-stable warp release.
- Benchmarking ``wp.Texture2D`` sampling vs the current direct array reads.

## Related

- Discussed in PR #2560 review thread (Eric's perf benchmarks + design exchange).

**Comment by preist-nvidia** [comment]
Backlog for now - HF is not the highest priority.
